ambry-server/src/main/java/com/github/ambry/server/RepairRequestsSender.java

```diff
@@ -0,0 +1,237 @@
+  // this data node
+  private final DataNodeId nodeId;
+  // LOCAL_CONSISTENCY_TODO. May listen to the Helix to get updated partition lists.
+  private final ClusterParticipant clusterParticipant;
+  // The database which stores the RepairRequestRecord
+  private final RepairRequestsDb db;
+  // partition id list this data node gets assigned
+  private final List<Long> partitionIds;
+  // Hashmap from partition id to the replicas
+  private final Map<Long, ReplicaId> partition2Replicas;
+  // generator to generate the correlation ids.
+  private final AtomicInteger correlationIdGenerator = new AtomicInteger(0);
+  // shutdown flag
+  private boolean shutdown = false;
+
+  /**
+   * RepairRequestsSender constructor
+   * @param requestChannel the channel to send requests and receive responses
+   * @param factory the {@link LocalNetworkClientFactory} used to generate the {@link LocalNetworkClient}
+   * @param clusterMap the cluster map
+   * @param nodeId this {@link DataNodeId}
+   * @param repairRequestsDb the database to get the RepairRequestRecord
+   * @param clusterParticipant {@link ClusterParticipant}
+   */
+  public RepairRequestsSender(RequestResponseChannel requestChannel, LocalNetworkClientFactory factory,
+      ClusterMap clusterMap, DataNodeId nodeId, RepairRequestsDb repairRequestsDb,
+      ClusterParticipant clusterParticipant) {
+    this.requestChannel = requestChannel;
+    this.client = factory.getNetworkClient();
+    this.clusterMap = clusterMap;
+    this.nodeId = nodeId;
+    this.clusterParticipant = clusterParticipant;
+
+    this.db = repairRequestsDb;
+
+    // LOCAL_CONSISTENCY_TODO. listen to the clustermap change and make the modification accordingly.
+    partitionIds = new ArrayList<>();
+    partition2Replicas = new HashMap<>();
+    for (ReplicaId replica : clusterMap.getReplicaIds(nodeId)) {
+      long partition = replica.getPartitionId().getId();
+      partitionIds.add(partition);
+      partition2Replicas.put(partition, replica);
+    }
+  }
+
+  public void run() {
+    List<RequestInfo> requestInfos = null;
+
+    while (!shutdown) {
+      try {
+        // sleep for some time
+        Thread.sleep(SLEEP_TIME_MS);
+
+        // loop all the partitions
+        for (long partitionId : partitionIds) {
+          if (shutdown) {
+            return;
+          }
+
+          // get the repair requests for this partition
+          requestInfos = getAmbryRequest(partitionId);
+          if (requestInfos == null || requestInfos.isEmpty()) {
+            continue;
+          }
+
+          // send the repair requests to the handler and wait for all the responses.
+          Set<Integer> requestsToDrop = new HashSet<>();
+          int inflightReqCount = requestInfos.size();
+          logger.info("RepairRequests Sender: {} sending {} repair requests ", nodeId, inflightReqCount);
+          List<ResponseInfo> responses = client.sendAndPoll(requestInfos, requestsToDrop, POLL_TIMEOUT_MS);
```

Comment:
I have some concerns over this logic. We are getting up to 100 (by default) repair requests from database, and the poll time is 20second. Since we only have one request handler thread for repair requests, each repair request would be handled in a sequence, which means, each request only has 200ms. when we have a busy server, this might be not enough time. we might need a for loop around client.sendAndPoll to make sure we have the same amount of responses coming back as the requests going out.

Reply:
You are right. Changed the logic to send request one by one. Thanks,